sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -903,15 +910,55 @@ private[joins] object LongHashedRelation {
       if (!rowKey.isNullAt(0)) {
         val key = rowKey.getLong(0)
         map.append(key, unsafeRow)
+      } else if (isNullAware) {
+        return new EmptyHashedRelationWithAllNullKeys
```

Comment:
Can variants of `EmptyHashedRelation` be singleton objects ? It can replace the instanceof check with an identity check

Reply:
do you mind giving me a code snippet, i am not sure I got your point.

Reply:
like
```
object EmptyHashedRelation extends ...
```

When we need to check it, just do `relation == EmptyHashedRelation`

Reply:
done